src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -1023,3 +1048,497 @@ address StubGenerator::generate_poly1305_processBlocks() {
+
+  poly1305_mul_reduce_vec4_avx2(YMM_ACC0, YMM_ACC1, YMM_ACC2,
+                          YMM_R0, YMM_R1, YMM_R2, YTMP1, YTMP2,
+                          YTMP3, YTMP4, YTMP5, YTMP6,
+                          YTMP7, YTMP8, YTMP9, t1);
+  __ vextracti128(YTMP1, YMM_ACC0, 1);
+  __ vextracti128(YTMP2, YMM_ACC1, 1);
+  __ vextracti128(YTMP3, YMM_ACC2, 1);
+
+  __ vpaddq(YMM_ACC0, YMM_ACC0, YTMP1, Assembler::AVX_128bit);
+  __ vpaddq(YMM_ACC1, YMM_ACC1, YTMP2, Assembler::AVX_128bit);
+  __ vpaddq(YMM_ACC2, YMM_ACC2, YTMP3, Assembler::AVX_128bit);
+
+  __ vpsrldq(YTMP1, YMM_ACC0, 8, Assembler::AVX_128bit);
+  __ vpsrldq(YTMP2, YMM_ACC1, 8, Assembler::AVX_128bit);
+  __ vpsrldq(YTMP3, YMM_ACC2, 8, Assembler::AVX_128bit);
+
+  // Finish folding and clear second qword
+  __ vpaddq(YMM_ACC0, YMM_ACC0, YTMP1, Assembler::AVX_128bit);
+  __ vpaddq(YMM_ACC1, YMM_ACC1, YTMP2, Assembler::AVX_128bit);
+  __ vpaddq(YMM_ACC2, YMM_ACC2, YTMP3, Assembler::AVX_128bit);
+
+  __ movq(YMM_ACC0, YMM_ACC0); //vmovq
+  __ movq(YMM_ACC1, YMM_ACC1); //vmovq
+  __ movq(YMM_ACC2, YMM_ACC2); //vmovq
+
+  // add     %%MSG, POLY1305_BLOCK_SIZE*4
+  // and     %%LEN, (POLY1305_BLOCK_SIZE*4 - 1) ; Get remaining lengths (LEN < 64 bytes)
+  __ lea(input, Address(input,16*4));
+  __ andq(length, 63+0); // remaining bytes < length 64
+  // carry propagation
+  __ vpsrlq(YTMP1, YMM_ACC0, 44, Assembler::AVX_128bit);
+  __ vpand(YMM_ACC0, YMM_ACC0, ExternalAddress(poly1305_mask44()), Assembler::AVX_128bit, t1); // Clear top 20 bits
+  __ vpaddq(YMM_ACC1, YMM_ACC1, YTMP1, Assembler::AVX_128bit);
+  __ vpsrlq(YTMP1, YMM_ACC1, 44, Assembler::AVX_128bit);
+  __ vpand(YMM_ACC1, YMM_ACC1, ExternalAddress(poly1305_mask44()), Assembler::AVX_128bit, t1); // Clear top 20 bits
+  __ vpaddq(YMM_ACC2, YMM_ACC2, YTMP1, Assembler::AVX_128bit);
+  __ vpsrlq(YTMP1, YMM_ACC2, 42, Assembler::AVX_128bit);
+  __ vpand(YMM_ACC2, YMM_ACC2, ExternalAddress(poly1305_mask42()), Assembler::AVX_128bit, t1); // Clear top 20 bits
+  __ vpsllq(YTMP2, YTMP1, 2, Assembler::AVX_128bit);
+  __ vpaddq(YTMP1, YTMP1, YTMP2, Assembler::AVX_128bit);
+  __ vpaddq(YMM_ACC0, YMM_ACC0, YTMP1, Assembler::AVX_128bit);
+
+  // Put together A
+  __ movq(a0, YMM_ACC0);
+  __ movq(t0, YMM_ACC1);
+  __ movq(t1, t0);
+  __ shlq(t1, 44);
+  __ orq(a0, t1);
+  __ shrq(t0, 20);
+  __ movq(a2, YMM_ACC2);
+  __ movq(a1, a2);
+  __ shlq(a1, 24);
+  __ orq(a1, t0);
+  __ shrq(a2, 40);
+
+  // SAFE DATA (clear powers of R) TODO: switch to vmovdqa
+  __ vpxor(YTMP1, YTMP1, YTMP1, Assembler::AVX_256bit);
+  __ vmovdqu(Address(rsp, _r4_r1_save + 0), YTMP1);
+  __ vmovdqu(Address(rsp, _r4_r1_save + 32), YTMP1);
+  __ vmovdqu(Address(rsp, _r4_r1_save + 32*2), YTMP1);
+  __ vmovdqu(Address(rsp, _r4_save + 0), YTMP1);
+  __ vmovdqu(Address(rsp, _r4_save + 32), YTMP1);
+  __ vmovdqu(Address(rsp, _r4_save + 32*2), YTMP1);
+  __ vmovdqu(Address(rsp, _r4p_save), YTMP1);
+  __ vmovdqu(Address(rsp, _r4p_save + 32), YTMP1);
```

Comment:
Also need to wipe all registers (`vzeroall` should be sufficient I believe, since only have `xmm0-xmm15`):
```
__ vzeroall();
for (int offset = _r4_r1_save; offset <= _r4p_save + 32; offset += 32) {
    __ vmovdqu(Address(rsp, offset), xmm0);
}
```